Move pending ranges functions into effective_replication_map

bhalevy opened the pull request:
Currently, the pending ranges are kept in token_metadata in
per-keyspace maps.

However, the pending endpoints aren't really per keyspace
so we shouldn't calculate them over and over again per keyspace.
    
They actually depend on the replication strategy, its configuration and the token_metadata,
hence they should be kept per effective_replication_map and use the precalculated address_ranges.

This series moves the function into e_r_m, and convert them to coroutines.
Also, it uses the e_r_m factory to iterate over all e_r_m:s for calling has_pending_ranges for each.

The main benefit from doing this is reducing memory footprint for the pending ranges and respectively saving cpu cycles that were used for recalculating them for multiple keyspaces that now share a replication map.

Fixes #9667 

Test: unit(dev)
Dtest: update_cluster_layout_tests:TestLargeScaleCluster.add_50_nodes_test update_cluster_layout_tests:TestUpdateClusterLayout.{simple_add_node_1_test,add_new_node_while_add_new_table_after_bootstrapping_test,increment_decrement_counters_in_threads_nodes_restarted_test,iterative_add_3_node_decommission_3_node_rf_2_test,simple_add_new_node_while_adding_info_2_test,simple_kill_new_node_while_bootstrapping_with_parallel_writes_in_multidc_test,verify_latest_copy_decommission_node_test,verify_latest_copy_rebuild_node_test,verify_latest_copy_removenode_node_test,verify_latest_copy_replace_node_test,verify_latest_copy_add_node_test}

avikivity commented on log.hh:
```diff
@@ -46,4 +49,18 @@ inline void apply_settings(const settings& s) {
+
+template <typename... Args>
+sstring debug_format(logger& l, const char* fmt, Args... args) {
+    return log_format(l, log_level::debug, fmt, std::forward<Args>(args)...);
```
Not sure I understand the motivation. Also, see bc75e2c1d182ae8440333b50feb71bb183dbfe57

bhalevy replied:
The motivation is to conditionally format a message string only when the logger's debug log level is enabled.
The message string is prepared in several call sites and then passed to a common function that logs it, again only in debug level.

An alternative, if we want to normally save these cycles could be to pass a function that will generate the message and call it only if debug log level is enabled, or do the debug-logging at the caller, before calling the common function (`mutate_token_metadata`)

avikivity commented on locator/abstract_replication_strategy.hh:
```diff
@@ -160,10 +164,15 @@ public:
     token_metadata_ptr _tmptr;
     replication_map _replication_map;
     size_t _replication_factor;
+    pending_ranges_interval_map _pending_ranges_interval_map;
```
Idea: add a pointer to the effective replication map for the same nodes, without any pending ranges. Perhaps it can reduce double calculation of things that don't change.

This is not so important here, but will be important if we fold dead nodes into the effective replication map, since these can change often.

bhalevy replied:
What do you mean by "same nodes"?
Can you please give an example?

avikivity replied:
Let's say this is the relication map of {n1, n2, n3, n4, n5, pn6}. Add a pointer to the replication map of {n1, n2, n3, n4, n5}. The read qorums of the two are the same.

Similarly, consider the replication map of a cluster with a dead node: {n1, n2, dn3, n4, n5}. Its replication map will be mostly similar to {n1, n2, n3, n4, n5}, and when the node comes back alive it will revert back to it.

avikivity commented on locator/abstract_replication_strategy.hh:
```diff
@@ -394,7 +396,15 @@ public:
+    template <typename Func>
+    requires requires (Func func, effective_replication_map_ptr erm) {
+        { futurize_invoke(func, erm) } -> std::same_as<future<>>;
+    }
```
It's enough to write std::invocable<effective_replication_map_ptr> IMO. Writing code in the concept means that the reader has to parse the code and understand it, whereas the named concept says: give me a function that accepts an effective_replication_map_ptr.

btw, isn't it nicer to accept const effective_replication_map&? the callee probably doesn't care for the pointer.

avikivity replied:
(then it simplfies to a single line: `future<> do_for_each(std::invocable<...> auto func) ... {`)

bhalevy replied:
makes sense, will do in next version

bhalevy replied:
> btw, isn't it nicer to accept const effective_replication_map&? the callee probably doesn't care for the pointer.

It's nicer, but if the callee is async it needs to keep the shard pointer to extend its life time.
It can call map.shared_from_this() if needed. I'll document that for future users.

bhalevy replied:
Hmm, this doesn't work well:
```
    template <typename Func>
    future<> do_for_each(std::invocable<Func, const effective_replication_map&> auto func) const noexcept {
```

```
service/storage_service.cc:1872:34: error: no matching member function for call to 'do_for_each'
            ss.get_erm_factory().do_for_each([endpoint] (const locator::effective_replication_map& erm) {
            ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
./locator/abstract_replication_strategy.hh:407:14: note: candidate template ignored: couldn't infer template argument 'Func'
    future<> do_for_each(std::invocable<Func, const effective_replication_map&> auto func) const noexcept {
             ^
```

while this does work:
```
    template <typename Func>
    requires std::invocable<Func, const effective_replication_map&>
    future<> do_for_each(Func func) const noexcept {
```

avikivity replied:
Drop `template <typename Func>`

avikivity replied:
All three lines drop to one:

```c++
future<> do_for_each(std::invocable<const effective_replication_map&> auto func) const noexcept {
```

bhalevy replied:
Fixed in v4

avikivity commented:
Looks reasonable, I didn't verify the copied code was unchanged, I assume it's exactly the same.

Perhaps we should move over from pending ranges to separate read quorums and write quorums, but that's for much later.

bhalevy commented:
> Looks reasonable, I didn't verify the copied code was unchanged, I assume it's exactly the same.

The algorithm was not changes, but I did change the implementation and improved it a bit when moving the code from token_metadata to effective_replication_map, and also used coroutines where previously the code required running in a seastar thread for yielding.

> 
> Perhaps we should move over from pending ranges to separate read quorums and write quorums, but that's for much later.

I agree it's a good direction. For writes we need to a quorum (or |CL| nodes in general) both in the current replica set and in the new one while it's changing, since node operations are abortable until committed in full.
This would be very important when adding or removing multiple nodes in a batch.

bhalevy commented:
In v2:
- dropped `debug_format`
  - just call slogger.debug before calls to mutate_token_metadata
- simplified effective_replication_map_factory::do_for_each template
  - using `requires std::invocable<Func, const effective_replication_map&>`
- rebased + retested

avikivity commented on locator/abstract_replication_strategy.hh:
```diff
@@ -399,6 +401,15 @@ public:
+    // If it yields, it must keep erm.shared_from_this() to extend its lifetime
+    template <typename Func>
+    requires std::invocable<Func, const effective_replication_map&>
+    future<> do_for_each(Func func) const noexcept {
```
Can simplify to `future<> do_for_each(std::invocable<const effective_replication_map&> to func)`

bhalevy replied:
Didn't work for me... Please see https://github.com/scylladb/scylla/pull/9703#discussion_r758967952

bhalevy replied:
Fixed in v4

avikivity commented:
> > Looks reasonable, I didn't verify the copied code was unchanged, I assume it's exactly the same.
> 
> The algorithm was not changes, but I did change the implementation and improved it a bit when moving the code from token_metadata to effective_replication_map, and also used coroutines where previously the code required running in a seastar thread for yielding.

Hmm, do we have tests for everything to ensure nothing was changed during the transformation?

> 
> > Perhaps we should move over from pending ranges to separate read quorums and write quorums, but that's for much later.
> 
> I agree it's a good direction. For writes we need to a quorum (or |CL| nodes in general) both in the current replica set and in the new one while it's changing, since node operations are abortable until committed in full. This would be very important when adding or removing multiple nodes in a batch.

If we merge dead nodes into the replication map, we could pre-calculate everything.

bhalevy commented:
(Fixed inadvertent Subproject commits)

bhalevy commented:
Passes all update_cluster_layout_tests dtests, except for simple_add_new_node_while_adding_info_1_test that fails on #8995 (reproduced same issue on master without this series)

bhalevy commented:
v3:
- followed up with a fix for #9667

avikivity commented on locator/abstract_replication_strategy.cc:
```diff
@@ -241,7 +241,7 @@ abstract_replication_strategy::get_address_ranges(const token_metadata& tm, inet
             rslogger.debug("token={} primary_range={} endpoint={}", t, r, endpoint);
             for (auto&& rng : r) {
-                ret.emplace(ep, rng);
+                ret.push_back(rng);
```
The change looks reasonable, but I don't see how the existing code caused a problem. How many address ranges can an endpoint participate in? At most 256*rf.

Unless it's everywhere_replication_strategy, in which case there are 256*nodes vnodes that are relevant to an endpoint. Still, even that doesn't explain how the multimap blew up.

bhalevy replied:
@avikivity it looks like the multimap blows up since we don't use an optimized structure for the token range.

I ran a small sanity test as follows and it it shows it's in the ball park:

```c++
#include <unordered_map>
#include <seastar/net/inet_address.hh>
#include <seastar/core/memory.hh>

SEASTAR_TEST_CASE(test_unordered_multimap) {
    enum class token_kind {
        before_all_keys,
        key,
        after_all_keys,
    };
    struct token {
        token_kind kind;
        int64_t data;
    };
    struct interval_bound {
        token value;
        bool inclusive;
    };
    struct token_range {
        std::optional<interval_bound> start;
        std::optional<interval_bound> end;
        bool singular;
    };
    std::unordered_multimap<net::inet_address, token_range> mm;
    int64_t tokens = 256;
    int nodes = 50;

    auto free_before = memory::stats().free_memory();

    for (auto node = 0; node < nodes; node++) {
        auto ia = net::inet_address(fmt::format("127.0.0.{}", node));
        for (int64_t t = 0; t < tokens; t++) {
            mm.emplace(ia, token_range{
                    interval_bound{token{token_kind::key, t}, true},
                    interval_bound{token{token_kind::key, t + 1}, false},
                    true});
        }
    }

    auto free_after = memory::stats().free_memory();

    fmt::print("consumed {} bytes", free_before - free_after);
    return make_ready_future<>();
}
```

```
Running 1 test case...
WARN  2021-12-05 11:40:35,498 [shard 0] seastar - Creation of perf_event based stall detector failed, falling back to posix timer: std::system_error (error system:13, perf_event_open() failed: Permission denied)
random-seed=2728524117
consumed 3031040 bytes
```

bhalevy replied:
Sorry, this doesn't necessarily explains the size of the "buckets" vector, but the overall memory consumption.

bhalevy replied:
Moving discussion to https://github.com/scylladb/scylla/issues/9667

bhalevy commented:
In v4:
- simplified `effective_replication_map_factory::do_for_each` using `std::invocable`
- abstract_replication_strategy: get_range_addresses: optimize for symmetric replication strategies
- rebased + retested

bhalevy commented:
I'm testing a couple more changes, will push again when tests pass.

bhalevy commented:
In v5:
- dropped "abstract_replication_strategy: get_range_addresses: optimize for symmetric replication strategies" and related patches
  - there are subtleties around get_range_addresses call sites that should be carefully examined. I'm not sure all of them can work with a single range.  Probably need to return a smarter object. maybe holding an interval map that can be searched to map a given token range.
  - Anyhow, I think we should consider that as a follow-up.
- "abstract_replication_strategy: get_address_ranges: return token_range_vector"
  - fixed typo in commit message
  - get_pending_address_ranges: no need to copy the returned vector
- Added:
  - "effective_repication_map: clear _pending_ranges_interval_map gently also from set_pending_ranges"
  - "effective_repication_map: change pending_ranges to unordered_map<range<token>, std::unordered_set<inet_address>"
  - "effective_replication_map: pre-calculate pending_nodes"
  - "effective_replication_map: pending_ranges_interval_map: keep inet_address_vector_topology_change"
- Rebased + retested

bhalevy commented:
@avikivity, please re-review

avikivity commented on db/view/view.cc:
```diff
@@ -1282,7 +1282,10 @@ future<> mutate_MV(
         auto target_endpoint = get_view_natural_endpoint(keyspace_name, base_token, view_token);
-        auto remote_endpoints = service::get_local_storage_proxy().get_token_metadata_ptr()->pending_endpoints_for(view_token, keyspace_name);
+        const auto& db = service::get_local_storage_proxy().local_db();
+        const auto& ks = db.find_keyspace(keyspace_name);
```
Idea: give storage_proxy its own keyspace type and storage and avoid the need to access database for replication information.

bhalevy replied:
makes sense.  need to see how this work out wit respect to the mechanics of creating such a keyspace.
(and out of scope of this PR :)

avikivity commented on locator/abstract_replication_strategy.cc:
```diff
@@ -341,10 +343,195 @@ inet_address_vector_replica_set effective_replication_map::get_natural_endpoints
+future<> effective_replication_map::calculate_pending_ranges_for_bootstrap(
+        pending_ranges& new_pending_ranges,
+        token_metadata& all_left_metadata,
+        const std::unordered_map<token, inet_address>& bootstrap_tokens) const {
```
What happens if it fails? Does it leave the inputs in some intermediate state?

bhalevy replied:
`bootstrap_tokens` is not changed.
`new_pending_ranges` and `all_left_metadata` may be left in an intermediate state but they are outputs allocated on the caller `effective_replication_map::update_pending_ranges()` stack anyhow.

bhalevy commented:
In v6:
- rebased + retested
- added "abstract_replication_strategy: get_range_addresses: return range_addresses object" (and some preparatory patches)
  - Optimized for symmetric replication strategies (like everywhere or local strategy) that use the same replica-set for all token ranges.

bhalevy commented:
This PR aged out and with the nove to consistent-topology-changes and tablets it's not required anymore
